### DISCUSSION TITLE
Bitcoin Dispute Resolution Module

### DISCUSSION THREAD

**Discussion by BitcoinBrief** [fedimint/fedimint#1614] [Fedimint Modules]
I have an idea for a Bitcoin-native, voluntary online dispute resolution (ODR) platform that solves the problem of enforcement. 
With very few exceptions, to be enforceable, decisions produced through ODR or alternative dispute resolution (ADR) systems must rely on the state-run court system and its monopoly on violence.  The exceptions, truly voluntary, opt-in systems of private enforcement, control the flow of funds in a closed system and are thus able to execute decisions by debiting the liable party's account and crediting the injured party -- eBay/paypal, is the archetypal example. 

And because bitcoin is the most seizure- and censorship-resistant money in human history, a hyperbitcoinized community will struggle to resolve civil disputes through monetary judgments without some clear system in place, to which all community members have consented, that renders bitcoin enforceable (subject to well-defined and fair adjudication). 

A bitcoin dispute resolution (₿DR) system will thus solve the traditional limitation of private dispute resolution (reliance on state-run courts for enforcement) and the challenges to the civil justice system posed by hyperbitcoinization.  

I think anchoring the ₿DR system into a FediMint module could provide the ability to automatically enforce decisions produced through the system as between mint members.  Only the enforcement logic would need to be at the module level.  This would allow many different flavors of dispute resolution systems to be plugged into the module as the oracle, depending on what adjudicatory system the community preferred.  

I have completed a working draft of a paper exploring this idea in detail, which will be peer reviewed by academics and eventually published with the Satoshi Papers.  But I'd prefer to have actual bitcoin builder input as well, not just arm chair academics, who may not understand bitcoin at all.  Any comments and guidance the fedimint community has is welcome!  A link to a google document of the working paper is here:
https://docs.google.com/document/d/1nWI5HPEzTq_6z1gNd1LkOfcjF8tp-W4k/edit?usp=sharing&ouid=107890407694920487238&rtpof=true&sd=true

I'm coming at this as a practicing attorney who is fed up with the fiat justice system and is trying to create viable alternatives for the burgeoning bitcoin economies.  Obi has aptly analogized bitcoin communities to frontier towns: https://twitter.com/obi/status/1602719065394626561.  Frontier towns in the American west either lacked rule of law, or were tyrannized by swingeing judges with little oversight.  ₿DR systems that are properly tailored to the individual communities could provide the enforceability necessary for rule of law, while affording ample due process -- especially if kept entirely opt-in and voluntary as a condition of joining a FediMint.  

Please let me know if any of this sounds viable and workable from a technical perspective.  Perhaps there is a more suitable infrastructure (DLCs come to mind).  

Thanks!

-Aaron D.

**Comment by humansinstitute** [comment]
Hey - this sounds kinda interesting but I'm not quite sure I understand the context fully.

Would this involve posting bonds against specific contracts that are registered in the fedimint and then having the guardians act as "defacto judges" that can mediate disputes or would you see that being achieved purely in code?

**Reply by BitcoinBrief** [reply]
> Would this involve posting bonds against specific contracts

I think enforcement can be achieved within the fedimint without using bonds, if the dispute is between fedimint community members who already have deposits in the fedimint.  This allows resolution of disputes for which there could be no bilateral ex-ante contractual agreement between the disputants that mandates bonds.  But it would require a multilateral contract that each fedimint depositor signed, binding that depositor to the bitcoin dispute resolution (BDR) protocol and authorizing/assenting/acknowledging that final decisions of the BDR process will result in automatic transfer of funds to enforce the decision. 

For example, say a community called Finneyland is hyperbitcoinized and most of its residents are depositors in a fedimint.  Upon taking up residence and/or depositing into the fedimint, residents agree to terms of service that specify all disputes between fedimint members shall be resolved by BDR and will be automatically enforced once all appeals are exhausted.  Alice runs over Bob's dog with her motorscooter, injuring it badly.  Bob could initiate a claim against Alice, which would trigger the BDR process and ultimately render a decision that Alice owes Bob his veterinary fees, as well as whatever fees the BDR provider charges.  Alice chooses not to appeal, and the ecash funds are automatically debited from Alice's wallet and credited to Bob's.  Obviously in this example, Alice and Bob had no contractual relationship related to her dog, and could not anticipate that Bob's claim would arise.  A BDR system reliant on ex-ante bilateral contracts and bonds would not serve fedimint members in this example, which is essentially a property claim that would be resolved in traditional small claims courts.  

> having the guardians act as "defacto judges" that can mediate disputes or would you see that being achieved purely in code?

The architecture of the design contemplates a separation between the fedimint module acting as an enforcement layer, and the BDR service provider, acting as the resolution layer.  This separation, among other things, would allow optionality in the dispute resolution mechanism chosen by the fedimint.  The fedimint module would primarily execute the decisions rendered by the separate BDR provider.  The BDR provider acts as the oracle that transmits the information to the fedimint module necessary to execute the decision (by transferring ecash from one member to another).  BDR providers thus could create platforms for fedimint communities to use that designate guardians as the neutral decision-makers, or other random fedimint members as a "jury", or the BDR providers could use totally separate professional mediators/arbitrators.  This dual layer architecture allows optionality at the resolution layer to suit the community.

The fedimint module would also need a way to alert the BDR system to a new claim or dispute, and transmit and receive the bare minimum information necessary for execution as between the disputants.  But the separate BDR provider would be the platform where disputants submit all their evidence and make their cases.  Thus, the dual layer architecture also minimizes the information disclosed to the fedimint through the adjudication process.

**Comment by jkitman** [comment]
Federated adjudication is an excellent idea, the need to not trust a single judge is why we form juries and councils to deliberate.  Multiple individuals are harder to fool or corrupt in a system of justice.

Your idea sounds like a federated oracle, no?  You could use a federation to come to consensus on the state of a contract.  In the case of objective world events (e.g. the price of BTC/USD) the federation could blindly settle the outcome of a contract if the parties use DLCs.  You could even institute a punishment mechanism for oracles that are too far out of consensus on "objective" truths.

However your description suggests the federation could also resolve civil cases that require specific adjudication.  In this case the parties would submit evidence out-of-band so that the federation could award/punish either party depending on the outcome.  The challenge here is how do you ensure the parties already have assets in the federation, especially since by default user balances are private?  In most civil disputes, one party is suing the other, and the federation has no coercive power over the defendant.

**Reply by BitcoinBrief** [reply]
> how do you ensure the parties already have assets in the federation

The fedimint members would agree to the bitcoin dispute resolution process at the point of depositing into the fedimint.  At scale, this could resemble the agreement residents of the free city Prospera sign upon taking up residence, agreeing to submit all disputes between other residents and the city operator to the dispute resolution forum designated in the agreement (here, the BDR protocol).  

Or, perhaps an industry group wants to create a private dispute system to adjudicate recurring disputes amongst themselves.  This exists with the auto-insurance industry in Israel, where all insurers have agreed to private adjudication outside the court system for fender-benders, and decisions are automatically enforced through a clearing-house system that has access to all the insurers' accounts.  The industry could create a fedimint and deposit their claims-related funds, agreeing in advance to the BDR system.

So, in either example, if someone initiates a claim against someone else, it's because they are both fedimint members who already have deposits in the mint.  I encourage everyone who is interested to skim through my paper, where I examine some of these use-cases and contexts.  It's linked [in my original post](https://github.com/fedimint/fedimint/discussions/1614#discussion-4833598) and I'll attach it again to this post. 

> especially since by default user balances are private? In most civil disputes, one party is suing the other, and the federation has no coercive power over the defendant.

User balances are private to the guardians and other users/depositors, but the user's wallets display the full balance.  There would have to be a way to unveil that total balance to the dispute resolution provider, as well, once a dispute reaches the level of binding adjudication.  This is a privacy trade-off, but one that would be necessary if the fedimint members need/want some level of rule of law in their community.  NOT ALL FEDIMINTS WILL MAKE THAT TRADE-OFF.  It's up the the members of the communities to determine that for themselves.  


[BDR - Bitcoin Dispute Resolution - (working draft).pdf](https://github.com/fedimint/fedimint/files/10737285/BDR.-.Bitcoin.Dispute.Resolution.-.working.draft.pdf)

**Comment by BitcoinBrief** [comment]
I've designed a dispute resolution system for adjudicating FOSS bounties that could possibly be deployed as a module.  The full paper is here:  https://github.com/BitResolve/Bounty-Adjudication-System

In the system, a federation could be used to escrow the bounty funds.  A review panel of third-party neutral developers considers the bounty submission and votes on whether to release the bounty to the applicant.  This review panel could be installed as a module to the federation.

Does fedimint make sense in this system?  It is appealing to me because keeping escrow transactions off-chain will minimize fees and allow for smaller bounties to be reviewed and adjudicated.  But a federated escrow means the guardians are jointly custodying the bounty funds.  This joint-custody arrangement introduces regulatory overhead/risk.  A fully noncustodial escrow (DLC, miniscript) would eliminate regulatory burdens.  

Interested in the community's thoughts!  

![2 2 - Bounty Adjudication Process](https://github.com/fedimint/fedimint/assets/115094037/3093aab0-8562-4c2a-8397-cbfd15db1e64)

**Reply by humansinstitute** [reply]
Hey - I'd originally replied on Nostr but I'm aware that's not always the best channel for finding things after the fact!

For me a benefit of implementing the escrow inside the fedimint as opposed to a DLC would be to simplify the payout structure. 

To lock and hold the funds in a DLC you would need to know the potential winners of the bounty when it is initially set, which you wouldn't know. I think an alternative option would be to implement this with a Threshold Escrow Contract inside Fedimint that would allow you to send the funds to any address as appropriate once the review panel have voted. Would also provide options to payout over lightning and with lower fees. 

An alternative to this, if you wanted to stay on chain would be to pay the bounty into multisig or perhaps even a Frost key. given the integration with Nostr there might be an interesting option to use something like [Frostr](https://github.com/nickfarrow/frostr) to control the funds that way the bounty holder could have a public presence on Nostr as well? 

https://snort.social/e/nevent1qqs9lr39gwzt2hx3fhuw2v3jhgg0twlj3pehzrvxa6c07a8mtf79w2c9u8uyk

**Reply by jkitman** [reply]
Could be an option for an organization such as the board of https://opensats.org/

Perhaps some of the grantees would prefer to receive via ecash for greater privacy.  Recurring payments could also be implemented, Spiral for instance does monthly payments which could be fiat-denominated.

If I were you I'd find an organization interested in experimenting with the first prototypes of such a module.

**Comment by tvolk131** [comment]
I love the dispute resolution idea! Here are some of my thoughts on how a Fedimint module could be designed:

When resolving disputes, the federation needs to be able to link a person's real-world identity to their federation funds in order to know who to debit. Further, in order for disputes to be enforced, the ₿DR needs to be able to move the losing party's funds without their immediate consent (that is, they initially agreed to put their funds in custody of the federation and agreed to the "Resident Agreement of Coexistence", but may not agree with the ₿DR's ruling for a particular dispute). This means we likely need users to place a subset of their funds into a timelock to prevent them from moving their funds out of a federation if they think a dispute will be raised against them. We'll call these KYC+timelocked funds "dispute collateral". I don't see any reason why dispute collateral couldn't be shared by a group, as long as that group is decided upon when staking the dispute collateral (a family could share a single bucket of dispute collateral if they agree to resolve internal family disputes on their own).

Next there is the issue of how to incentivize users to add dispute collateral. Dispute collateral comes with two fairly considerable downsides as mentioned earlier:

1. KYC (since the federation needs to be able to debit you if a dispute is raised against you)
2. Timelock (so you don't flee the federation to avoid a dispute)

The simplest solution here would be to require a minimum amount of dispute collateral in the "Resident Agreement of Coexistence" as a prerequisite to being a resident of whatever community the federation is running. However, ideally we could incentivize users to add enough dispute collateral to be able to resolve a high ratio of potential disputes, and the benefits of staking outweigh the two downsides above. I would propose that, by increasing your dispute collateral, you increase the amount you can win in a dispute. Let's take the situation where Alice hits Bob's dog with her motor scooter and refuses to pay for the incurred veterinarian bills. If Bob's veterinarian bills are 1 million sats, but he only has 500k sats of dispute collateral staked in his name, then even if Alice has 1 million sats of dispute collateral (which would mean that the federation has the _enforcement power_ to reimburse Bob for the full cost), he would only receive 500k sats, meaning Alice would only have to pay for half of the cost. This incentivizes well-behaved users to add dispute collateral since they may be rewarded more from disputes, and will not be punished disproportionately compared to those who stake less in dispute collateral. So let's say the federation debits Alice 500k sats and pays them to Bob. Afterwards, Alice continues to have disputes filed against her. As her dispute collateral continues to be debited, two things happen:

1. The amount she can win from raising disputes gets smaller and smaller, eventually reaching zero which effectively blocks her from raising any disputes
2. If the "Resident Agreement of Coexistence" states a minimum amount of dispute collateral, any real-world penalties stated in the contract could take effect (but this is obviously outside the scope of a Fedimint module)

This design would help align incentives toward cooperation and allow users to choose how to split their funds between private spendable E-Cash and KYC'd dispute collateral.

**Reply by BitcoinBrief** [reply]
Thanks for the well-reasoned thoughts! Your idea is to essentially cap a party's damages in any one dispute to the amount of dispute collateral in time lock. This is a simple solution to the problem of incentivizing federation users/community members to keep funds available for dispute. 

We could also tweak the cap so it's not 1:1, but a sliding scale, depending on the circumstances. 

Currently, a team is building a dispute resolution fedimint module for open source bounties in the [Bolt.Fun hackathon](https://bolt.fun/project/resolvr), called Resolvr.  Please check it out at this repo: https://github.com/Resolvr-io. And pop into the [discord](https://discord.com/invite/VRmwgV3pjP) for real time discussion.  We need all the input we can get!

@m1sterc001guy is building FROST into fedimint to handle external nostr and bitcoin signing events. Some exciting stuff happening!

-Aaron

**Reply by tvolk131** [reply]
> We could also tweak the cap so it's not 1:1, but a sliding scale, depending on the circumstances.

Very good point! If a dispute involves not just accidental damage of property, but is clearly intentional or even premeditated, the cap could be greater than 1:1 without skewing dispute collateral incentives for participants who have no intent on committing intentional or premeditated crimes. Fortunately, all of this complexity regarding categories of disputes with different dispute collateral ratios could be handled outside of a Fedimint module (it could be specified in plain English in the Resident Agreement of Coexistence). All the module would need to handle is the federated decisions to move dispute collateral from one real-world entity to another, and rely on the federated decision making process to ensure that the dispute collateral ratios specified in the Resident Agreement of Coexistence are followed.

**Comment by EthnTuttle** [comment]
FWIW, I think BitVM has potential in this realm. Will try to gather more concrete thoughts on the matter, but I'm still grokking the whole thing now.
